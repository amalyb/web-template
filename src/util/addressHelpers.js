/**
 * Address normalization and mapping utilities
 */

// Normalize address data by trimming spaces and converting to standard formats
export const normalizeAddress = (address) => {
  if (!address) return null;

  const normalized = {
    fullName: address.fullName?.trim().replace(/\s+/g, ' ') || '',
    country: address.country?.toUpperCase() || 'US',
    street1: address.street1?.trim().replace(/\s+/g, ' ') || '',
    street2: address.street2?.trim().replace(/\s+/g, ' ') || undefined, // Always send undefined when missing
    city: address.city?.trim().replace(/\s+/g, ' ') || '',
    state: address.state?.trim().toUpperCase() || '',
    zip: address.zip?.trim().replace(/\s+/g, '') || '',
    email: address.email?.trim().toLowerCase() || '',
    phone: normalizePhone(address.phone) || '',
  };

  // Clean up empty strings to undefined for optional fields
  if (!normalized.street2) normalized.street2 = undefined;
  if (!normalized.phone) normalized.phone = undefined;

  return normalized;
};

// Normalize phone number to E.164 format
export const normalizePhone = (phone) => {
  if (!phone) return '';
  
  // Remove extensions (x123, ext 123, etc.) and reject them
  let cleaned = phone.replace(/\s*(x|ext|extension)\s*\d+.*$/i, '');
  
  // Remove all non-digit characters except +
  cleaned = cleaned.replace(/[^\d+]/g, '');
  
  // If it doesn't start with +, assume US number and add +1
  if (!cleaned.startsWith('+')) {
    // Remove leading 1 if present
    if (cleaned.startsWith('1') && cleaned.length === 11) {
      cleaned = cleaned.substring(1);
    }
    cleaned = '+1' + cleaned;
  }
  
  // Validate E.164 format (1-15 digits after +)
  const e164Regex = /^\+[1-9]\d{1,14}$/;
  if (!e164Regex.test(cleaned)) {
    return phone; // Return original if normalization fails
  }
  
  return cleaned;
};

// Map normalized address to Stripe billing format
export const mapToStripeBilling = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    address: {
      line1: normalized.street1,
      line2: normalized.street2 || undefined, // Always send undefined when missing
      city: normalized.city,
      state: normalized.state,
      postal_code: normalized.zip,
      country: normalized.country,
    },
  };
};

// Map normalized address to Shippo/UPS/FedEx format
export const mapToShippo = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    street1: normalized.street1,
    street2: normalized.street2 || '',
    city: normalized.city,
    state: normalized.state,
    zip: normalized.zip,
    country: normalized.country,
  };
};

// Check if address is a PO Box (for courier routing)
export const isPOBox = (address) => {
  if (!address?.street1) return false;
  return /^(P(OST)?\.?\s*O(FFICE)?\.?\s*BOX)\b/i.test(address.street1);
};

// Map normalized address to UPS format
export const mapToUPS = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    address_line1: normalized.street1,
    address_line2: normalized.street2 || '',
    city: normalized.city,
    state_province_code: normalized.state,
    postal_code: normalized.zip,
    country_code: normalized.country,
  };
};

// Map normalized address to FedEx format
export const mapToFedEx = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    contact: {
      person_name: normalized.fullName,
      email_address: normalized.email,
      phone_number: normalized.phone,
    },
    address: {
      street_lines: [normalized.street1, normalized.street2].filter(Boolean),
      city: normalized.city,
      state_or_province_code: normalized.state,
      postal_code: normalized.zip,
      country_code: normalized.country,
    },
  };
};

// Validate address completeness
export const validateAddress = (address, requiredFields = {}) => {
  const defaults = {
    fullName: true,
    country: true,
    street1: true,
    city: true,
    state: true,
    zip: true,
    email: true,
    phone: false,
  };
  
  const requirements = { ...defaults, ...requiredFields };
  const errors = {};
  
  if (requirements.fullName && !address.fullName?.trim()) {
    errors.fullName = 'Full name is required';
  }
  
  if (requirements.country && !address.country?.trim()) {
    errors.country = 'Country is required';
  }
  
  if (requirements.street1 && !address.street1?.trim()) {
    errors.street1 = 'Street address is required';
  }
  
  if (requirements.city && !address.city?.trim()) {
    errors.city = 'City is required';
  }
  
  if (requirements.state && !address.state?.trim()) {
    errors.state = 'State/Region is required';
  }
  
  if (requirements.zip && !address.zip?.trim()) {
    errors.zip = 'ZIP/Postal code is required';
  }
  
  if (requirements.email && !address.email?.trim()) {
    errors.email = 'Email is required';
  } else if (address.email && !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(address.email)) {
    errors.email = 'Please enter a valid email address';
  }
  
  if (requirements.phone && !address.phone?.trim()) {
    errors.phone = 'Phone number is required';
  }
  
  return {
    isValid: Object.keys(errors).length === 0,
    errors,
  };
};

// Copy billing address to shipping address
export const copyBillingToShipping = (billingAddress) => {
  const normalized = normalizeAddress(billingAddress);
  if (!normalized) return null;

  return {
    shipping: {
      fullName: normalized.fullName,
      country: normalized.country,
      street1: normalized.street1,
      street2: normalized.street2,
      city: normalized.city,
      state: normalized.state,
      zip: normalized.zip,
      email: normalized.email,
      phone: normalized.phone,
    },
  };
};
