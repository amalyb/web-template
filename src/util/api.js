// These helpers are calling this template's own server-side routes
// so, they are not directly calling Marketplace API or Integration API.
// You can find these api endpoints from 'server/api/...' directory

import axios from 'axios';
import appSettings from '../config/settings';
import { types as sdkTypes, transit } from './sdkLoader';
import Decimal from 'decimal.js';

export const apiBaseUrl = marketplaceRootURL => {
  const port = process.env.REACT_APP_DEV_API_SERVER_PORT;
  const useDevApiServer = process.env.NODE_ENV === 'development' && !!port;

  // In development, the dev API server is running in a different port
  if (useDevApiServer) {
    return `http://localhost:${port}`;
  }

  // Otherwise, use the given marketplaceRootURL parameter or the same domain and port as the frontend
  return marketplaceRootURL ? marketplaceRootURL.replace(/\/$/, '') : `${window.location.origin}`;
};

// Axios client for API calls with Basic authentication
const apiBase = process.env.REACT_APP_API_BASE_URL || '/api';
export const apiClient = axios.create({ baseURL: apiBase });

apiClient.interceptors.request.use((config) => {
  const u = process.env.REACT_APP_BASIC_AUTH_USERNAME;
  const p = process.env.REACT_APP_BASIC_AUTH_PASSWORD;
  if (u && p) {
    const token = btoa(`${u}:${p}`);
    config.headers = { ...config.headers, Authorization: `Basic ${token}` };
  }
  return config;
});

// Application type handlers for JS SDK.
//
// NOTE: keep in sync with `typeHandlers` in `server/api-util/sdk.js`
export const typeHandlers = [
  // Use Decimal type instead of SDK's BigDecimal.
  {
    type: sdkTypes.BigDecimal,
    customType: Decimal,
    writer: v => new sdkTypes.BigDecimal(v.toString()),
    reader: v => new Decimal(v.value),
  },
];

const serialize = data => {
  return transit.write(data, { typeHandlers, verbose: appSettings.sdk.transitVerbose });
};

const deserialize = str => {
  return transit.read(str, { typeHandlers });
};

const methods = {
  POST: 'POST',
  GET: 'GET',
  PUT: 'PUT',
  PATCH: 'PATCH',
  DELETE: 'DELETE',
};

// If server/api returns data from SDK, you should set Content-Type to 'application/transit+json'
const request = (path, options = {}) => {
  const url = `${apiBaseUrl()}${path}`;
  const { credentials, headers, body, ...rest } = options;

  // If headers are not set, we assume that the body should be serialized as transit format.
  const shouldSerializeBody =
    (!headers || headers['Content-Type'] === 'application/transit+json') && body;
  const bodyMaybe = shouldSerializeBody ? { body: serialize(body) } : {};

  const fetchOptions = {
    credentials: credentials || 'include',
    // Since server/api mostly talks to Marketplace API using SDK,
    // we default to 'application/transit+json' as content type (as SDK uses transit).
    headers: headers || { 'Content-Type': 'application/transit+json' },
    ...bodyMaybe,
    ...rest,
  };

  return window.fetch(url, fetchOptions).then(res => {
    const contentTypeHeader = res.headers.get('Content-Type');
    const contentType = contentTypeHeader ? contentTypeHeader.split(';')[0] : null;

    if (res.status >= 400) {
      return res.json().then(data => {
        let e = new Error();
        e = Object.assign(e, data);

        throw e;
      });
    }
    if (contentType === 'application/transit+json') {
      return res.text().then(deserialize);
    } else if (contentType === 'application/json') {
      return res.json();
    }
    return res.text();
  });
};

// Keep the previous parameter order for the post method.
// For now, only POST has own specific function, but you can create more or use request directly.
export const post = (path, body, options = {}) => {
  const requestOptions = {
    ...options,
    method: methods.POST,
    body,
  };

  return request(path, requestOptions);
};

// Fetch transaction line items from the local API endpoint.
//
// See `server/api/transaction-line-items.js` to see what data should
// be sent in the body.
export const transactionLineItems = (payload) =>
  apiClient.post('/transaction-line-items', payload);

// Initiate a privileged transaction.
//
// With privileged transitions, the transactions need to be created
// from the backend. This endpoint enables sending the order data to
// the local backend, and passing that to the Marketplace API.
//
// See `server/api/initiate-privileged.js` to see what data should be
// sent in the body.
export const initiatePrivileged = body => {
  return post('/api/initiate-privileged', body);
};

// Transition a transaction with a privileged transition.
//
// This is similar to the `initiatePrivileged` above. It will use the
// backend for the transition. The backend endpoint will add the
// payment line items to the transition params.
//
// See `server/api/transition-privileged.js` to see what data should
// be sent in the body.
export const transitionPrivileged = body => {
  return post('/api/transition-privileged', body);
};

// Create user with identity provider (e.g. Facebook or Google)
//
// If loginWithIdp api call fails and user can't authenticate to Marketplace API with idp
// we will show option to create a new user with idp.
// For that user needs to confirm data fetched from the idp.
// After the confirmation, this endpoint is called to create a new user with confirmed data.
//
// See `server/api/auth/createUserWithIdp.js` to see what data should
// be sent in the body.
export const createUserWithIdp = body => {
  return post('/api/auth/create-user-with-idp', body);
};

// Ensure phone number is saved to protectedData
// This can be called after sign-up or profile updates to guarantee the phone number is stored correctly
export const ensurePhoneNumber = phoneNumber => {
  if (!phoneNumber) {
    console.warn('‚ö†Ô∏è [api] No phone number provided to ensurePhoneNumber');
    return Promise.resolve();
  }
  
  console.log('üì± [api] Ensuring phone number is saved to protectedData:', phoneNumber);
  return post('/api/ensure-phone-number', { phoneNumber })
    .then(() => {
      console.log('‚úÖ [api] Phone number ensured in protectedData');
    })
    .catch(error => {
      console.warn('‚ö†Ô∏è [api] Failed to ensure phone number:', error.message);
      throw error;
    });
};
