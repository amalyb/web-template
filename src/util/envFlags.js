/**
 * Centralized environment flags for client-side code
 * Safely handles missing process object in browser environments
 */

export const IS_PROD = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'production');
export const IS_DEV  = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'development');
export const IS_TEST = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'test');
export const __DEV__ = !IS_PROD;
