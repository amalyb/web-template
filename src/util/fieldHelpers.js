import { SCHEMA_TYPE_MULTI_ENUM, SCHEMA_TYPE_TEXT } from './types';

const keyMapping = {
  userType: {
    wrapper: 'userTypeConfig',
    limitTo: 'limitToUserTypeIds',
    ids: 'userTypeIds',
  },
  listingType: {
    wrapper: 'listingTypeConfig',
    limitTo: 'limitToListingTypeIds',
    ids: 'listingTypeIds',
  },
};

const getEntityTypeRestrictions = (entityTypeKey, config) => {
  const { wrapper, limitTo, ids } = keyMapping[entityTypeKey];
  const isLimited = config && config[wrapper] && config[wrapper][limitTo];
  const limitToIds = config && config[wrapper] && config[wrapper][ids];

  return { isLimited, limitToIds };
};

/**
 * Returns the value of the attribute in extended data.
 * @param {*} data extended data containing the value
 * @param {*} key attribute key in extended data
 * @returns
 */
export const getFieldValue = (data, key) => {
  const value = data?.[key];
  return value != null ? value : null;
};

/**
 * Pick props for SectionMultiEnumMaybe and SectionTextMaybe display components.
 * @param {*} publicData entity public data containing the value(s) to be displayed
 * @param {*} metadata entity metadata containing the value(s) to be displayed
 * @param {*} fieldConfig custom field configuration object
 * @param {*} entityTypeKey the name of the key denoting the entity type in publicData.
 * Currently supports 'userType' and 'listingType'.
 * @returns an object with attributes 'schemaType', 'key', and 'heading', as well as either
 * - 'options' and 'selectedOptions' for SCHEMA_TYPE_MULTI_ENUM
 * - or 'text' for SCHEMA_TYPE_TEXT
 *
 */
export const pickCustomFieldProps = (publicData, metadata, fieldConfig, entityTypeKey) => {
  return fieldConfig?.reduce((pickedElements, config) => {
    const { key, enumOptions, schemaType, showConfig = {}, scope = 'public' } = config;
    const { displayInProfile } = showConfig;
    const { isLimited, limitToIds } = getEntityTypeRestrictions(entityTypeKey, config);
    const entityType = publicData && publicData[entityTypeKey];
    const isTargetEntityType = !isLimited || limitToIds.includes(entityType);

    const createFilterOptions = options =>
      options.map(o => ({ key: `${o.option}`, label: o.label }));

    const value =
      scope === 'public'
        ? getFieldValue(publicData, key)
        : scope === 'metadata'
        ? getFieldValue(metadata, key)
        : null;

    return isTargetEntityType && schemaType === SCHEMA_TYPE_MULTI_ENUM && displayInProfile
      ? [
          ...pickedElements,
          {
            schemaType,
            key,
            heading: config?.showConfig?.label,
            options: createFilterOptions(enumOptions),
            selectedOptions: value || [],
          },
        ]
      : isTargetEntityType && !!value && config.schemaType === SCHEMA_TYPE_TEXT && displayInProfile
      ? [
          ...pickedElements,
          {
            schemaType,
            key,
            heading: config?.showConfig?.label,
            text: value,
          },
        ]
      : pickedElements;
  }, []);
};
