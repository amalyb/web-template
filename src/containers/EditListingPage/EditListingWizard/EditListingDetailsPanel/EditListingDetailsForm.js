import React from 'react';
import { arrayOf, bool, func, shape, string } from 'prop-types';
import { compose } from 'redux';
import { Field, Form as FinalForm } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import classNames from 'classnames';

// Import util modules
import { intlShape, injectIntl, FormattedMessage } from '../../../../util/reactIntl';
import { EXTENDED_DATA_SCHEMA_TYPES, propTypes } from '../../../../util/types';
import { maxLength, required, composeValidators } from '../../../../util/validators';

// Import shared components
import { Form, Button, FieldSelect, FieldTextInput } from '../../../../components';
// Import modules from this directory
import CustomExtendedDataField from '../CustomExtendedDataField';
import css from './EditListingDetailsForm.module.css';

const TITLE_MAX_LENGTH = 60;

// Show various error messages
const ErrorMessage = props => {
  const { fetchErrors } = props;
  const { updateListingError, createListingDraftError, showListingsError } = fetchErrors || {};
  const errorMessage = updateListingError ? (
    <FormattedMessage id="EditListingDetailsForm.updateFailed" />
  ) : createListingDraftError ? (
    <FormattedMessage id="EditListingDetailsForm.createListingDraftError" />
  ) : showListingsError ? (
    <FormattedMessage id="EditListingDetailsForm.showListingFailed" />
  ) : null;

  if (errorMessage) {
    return <p className={css.error}>{errorMessage}</p>;
  }
  return null;
};

// Hidden input field
const FieldHidden = props => {
  const { name } = props;
  return (
    <Field id={name} name={name} type="hidden" className={css.unitTypeHidden}>
      {fieldRenderProps => <input {...fieldRenderProps?.input} />}
    </Field>
  );
};

// Field component that either allows selecting transaction type (if multiple types are available)
// or just renders hidden fields:
// - transactionType
// - transactionProcessAlias
// - unitType
const FieldSelectTransactionType = props => {
  const {
    name,
    transactionTypes,
    hasExistingTransactionType,
    onProcessChange,
    formApi,
    intl,
  } = props;
  const hasMultipleTransactionTypes = transactionTypes?.length > 1;

  const handleOnChange = value => {
    const transactionProcessAlias = formApi.getFieldState('transactionProcessAlias')?.value;
    const selectedProcess = transactionTypes.find(config => config.transactionType === value);
    formApi.change('transactionProcessAlias', selectedProcess.transactionProcessAlias);
    formApi.change('unitType', selectedProcess.unitType);

    const hasProcessChanged = transactionProcessAlias === selectedProcess.transactionProcessAlias;
    if (onProcessChange && hasProcessChanged) {
      onProcessChange(selectedProcess.transactionProcessAlias);
    }
  };

  return hasMultipleTransactionTypes && !hasExistingTransactionType ? (
    <>
      <FieldSelect
        id={name}
        name={name}
        className={css.transactionTypeSelect}
        label={intl.formatMessage({ id: 'EditListingDetailsForm.transactionTypeLabel' })}
        validate={required(
          intl.formatMessage({ id: 'EditListingDetailsForm.transactionTypeRequired' })
        )}
        onChange={handleOnChange}
      >
        <option disabled value="">
          {intl.formatMessage({ id: 'EditListingDetailsForm.transactionTypePlaceholder' })}
        </option>
        {transactionTypes.map(config => {
          const type = config.transactionType;
          return (
            <option key={type} value={type}>
              {config.label}
            </option>
          );
        })}
      </FieldSelect>
      <FieldHidden name="transactionProcessAlias" />
      <FieldHidden name="unitType" />
    </>
  ) : hasMultipleTransactionTypes && hasExistingTransactionType ? (
    <div className={css.transactionTypeSelect}>
      <h5 className={css.selectedLabel}>
        {intl.formatMessage({ id: 'EditListingDetailsForm.transactionTypeLabel' })}
      </h5>
      <p className={css.selectedValue}>{formApi.getFieldState(name)?.value}</p>
      <FieldHidden name={name} />
      <FieldHidden name="transactionProcessAlias" />
      <FieldHidden name="unitType" />
    </div>
  ) : (
    <>
      <FieldHidden name={name} />
      <FieldHidden name="transactionProcessAlias" />
      <FieldHidden name="unitType" />
    </>
  );
};

// Add collect data for extended data fields (both publicData and privateData) based on configuration
const AddCustomExtendedDataFields = props => {
  const { transactionProcessAlias, listingExtendedDataConfig, intl } = props;
  const extendedDataConfigs = listingExtendedDataConfig || [];
  const fields = extendedDataConfigs.reduce((pickedFields, extendedDataConfig) => {
    const { key, includeForProcessAliases = [], schemaType, scope } = extendedDataConfig || {};

    const isKnownSchemaType = EXTENDED_DATA_SCHEMA_TYPES.includes(schemaType);
    const isTargetProcessAlias = includeForProcessAliases.includes(transactionProcessAlias);
    const isProviderScope = ['public', 'private'].includes(scope);

    return isKnownSchemaType && isTargetProcessAlias && isProviderScope
      ? [
          ...pickedFields,
          <CustomExtendedDataField
            key={key}
            name={key}
            fieldConfig={extendedDataConfig}
            defaultRequiredMessage={intl.formatMessage({
              id: 'EditListingDetailsForm.defaultRequiredMessage',
            })}
          />,
        ]
      : pickedFields;
  }, []);

  return <>{fields}</>;
};

// Form that asks title, description, transaction process and unit type for pricing
// In addition, it asks about custom fields according to marketplace-custom-config.js
const EditListingDetailsFormComponent = props => (
  <FinalForm
    {...props}
    mutators={{ ...arrayMutators }}
    render={formRenderProps => {
      const {
        autoFocus,
        className,
        disabled,
        ready,
        form: formApi,
        handleSubmit,
        onProcessChange,
        intl,
        invalid,
        pristine,
        selectableTransactionTypes,
        hasExistingTransactionType,
        saveActionMsg,
        updated,
        updateInProgress,
        fetchErrors,
        listingExtendedDataConfig,
        values,
      } = formRenderProps;

      const { transactionProcessAlias } = values;

      const titleRequiredMessage = intl.formatMessage({
        id: 'EditListingDetailsForm.titleRequired',
      });
      const maxLengthMessage = intl.formatMessage(
        { id: 'EditListingDetailsForm.maxLength' },
        {
          maxLength: TITLE_MAX_LENGTH,
        }
      );
      const maxLength60Message = maxLength(maxLengthMessage, TITLE_MAX_LENGTH);

      const classes = classNames(css.root, className);
      const submitReady = (updated && pristine) || ready;
      const submitInProgress = updateInProgress;
      const submitDisabled = invalid || disabled || submitInProgress;

      return (
        <Form className={classes} onSubmit={handleSubmit}>
          <ErrorMessage fetchErrors={fetchErrors} />

          <FieldTextInput
            id="title"
            name="title"
            className={css.title}
            type="text"
            label={intl.formatMessage({ id: 'EditListingDetailsForm.title' })}
            placeholder={intl.formatMessage({ id: 'EditListingDetailsForm.titlePlaceholder' })}
            maxLength={TITLE_MAX_LENGTH}
            validate={composeValidators(required(titleRequiredMessage), maxLength60Message)}
            autoFocus={autoFocus}
          />

          <FieldTextInput
            id="description"
            name="description"
            className={css.description}
            type="textarea"
            label={intl.formatMessage({ id: 'EditListingDetailsForm.description' })}
            placeholder={intl.formatMessage({
              id: 'EditListingDetailsForm.descriptionPlaceholder',
            })}
            validate={required(
              intl.formatMessage({
                id: 'EditListingDetailsForm.descriptionRequired',
              })
            )}
          />

          <FieldSelectTransactionType
            name="transactionType"
            transactionTypes={selectableTransactionTypes}
            hasExistingTransactionType={hasExistingTransactionType}
            onProcessChange={onProcessChange}
            formApi={formApi}
            intl={intl}
          />

          <AddCustomExtendedDataFields
            transactionProcessAlias={transactionProcessAlias}
            listingExtendedDataConfig={listingExtendedDataConfig}
            intl={intl}
          />

          <Button
            className={css.submitButton}
            type="submit"
            inProgress={submitInProgress}
            disabled={submitDisabled}
            ready={submitReady}
          >
            {saveActionMsg}
          </Button>
        </Form>
      );
    }}
  />
);

EditListingDetailsFormComponent.defaultProps = {
  className: null,
  fetchErrors: null,
  onProcessChange: null,
  hasExistingTransactionType: false,
  listingExtendedDataConfig: null,
};

EditListingDetailsFormComponent.propTypes = {
  className: string,
  intl: intlShape.isRequired,
  onSubmit: func.isRequired,
  onProcessChange: func,
  saveActionMsg: string.isRequired,
  disabled: bool.isRequired,
  ready: bool.isRequired,
  updated: bool.isRequired,
  updateInProgress: bool.isRequired,
  fetchErrors: shape({
    createListingDraftError: propTypes.error,
    showListingsError: propTypes.error,
    updateListingError: propTypes.error,
  }),
  selectableTransactionTypes: arrayOf(
    shape({
      transactionType: string.isRequired,
      transactionProcessAlias: string.isRequired,
      unitType: string.isRequired,
    })
  ).isRequired,
  hasExistingTransactionType: bool,
  listingExtendedDataConfig: propTypes.listingExtendedDataConfig,
};

export default compose(injectIntl)(EditListingDetailsFormComponent);
