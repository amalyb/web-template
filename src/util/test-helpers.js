import React from 'react';
import mapValues from 'lodash/mapValues';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import configureStore from '../store';
import { IntlProvider } from '../util/reactIntl';
import defaultConfig from '../config/defaultConfig';
import { ConfigurationProvider } from '../context/configurationContext';
import { RouteConfigurationProvider } from '../context/routeConfigurationContext';
import routeConfiguration from '../routing/routeConfiguration';

// In case you have translated the template and have new translations that
// are missing from the en translations file, the language for the tests can
// be changed here so that there are no missing translation keys in tests.
import messages from '../translations/en.json';

Enzyme.configure({ adapter: new Adapter() });

let undefined;
export const getDefaultConfiguration = () => {
  return {
    ...defaultConfig,
    currency: 'USD',
    facebookAppId: undefined,
    canonicalRootURL: 'http://localhost:3000',
    maps: {
      ...defaultConfig.maps,
      mapboxAccessToken: undefined,
      googleMapsAPIKey: undefined,
    },
    stripe: {
      ...defaultConfig.stripe,
      publishableKey: 'pk_test_',
    },
  };
};

export const getRouteConfiguration = () => {
  const layoutConfig = {
    searchPageVariant: 'map',
    listingPageVariant: 'full-image',
    listingImage: {
      aspectWidth: 400,
      aspectHeight: 400,
      variantPrefix: 'listing-card',
    },
  };
  return routeConfiguration(layoutConfig);
};

// Locale should not affect the tests. We ensure this by providing
// messages with the key as the value of each message.
const testMessages = mapValues(messages, (val, key) => key);

// Provide all the context for components that connect to the Redux
// store, i18n, router, etc.
export const TestProvider = props => {
  const store = configureStore();
  return (
    <ConfigurationProvider value={getDefaultConfiguration()}>
      <RouteConfigurationProvider value={getRouteConfiguration()}>
        <IntlProvider locale="en" messages={testMessages} textComponent="span">
          <BrowserRouter>
            <Provider store={store}>{props.children}</Provider>
          </BrowserRouter>
        </IntlProvider>
      </RouteConfigurationProvider>
    </ConfigurationProvider>
  );
};

// Use Enzyme's shallow rendering to render the given component to a
// JSON structure that can be used in snapshot tests. This doesn't
// render the children within the given component, only a
// representation of the child component and its props.
//
// Useful for snapshot testing components that contain shared
// components. With deep rendering, if the child component changes
// internally, the test for the given component would also fail. This
// avoids the problem by not rendering the full tree but only the
// relevant structure for the given component.
export const renderShallow = component => {
  return toJson(shallow(component));
};

// Fully render the given component to a JSON structure that can be
// used in snapshot tests.
export const renderDeep = component => {
  return toJson(mount(<TestProvider>{component}</TestProvider>), { mode: 'deep' });
};
