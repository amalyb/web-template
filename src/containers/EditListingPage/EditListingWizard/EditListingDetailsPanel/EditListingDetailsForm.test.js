import React from 'react';
import { renderDeep } from '../../../../util/test-helpers';
import { fakeIntl } from '../../../../util/test-data';
import EditListingDetailsForm from './EditListingDetailsForm';

const noop = () => null;
const processInfos = [
  {
    name: 'flex-product-default-process',
    alias: 'release-1',
    unitTypes: ['item'],
  },
];

const filters = [
  {
    id: 'category',
    label: 'Category',
    wizardPlaceholder: 'Chooseâ€¦',
    wizardRequired: 'You need to select a category.',
    type: 'SelectSingleFilter',
    group: 'primary',
    queryParamNames: ['pub_category'],
    config: {
      // Schema type is enum for SelectSingleFilter
      schemaType: 'enum',

      // "key" is the option you see in Flex Console.
      // "label" is set here for the UI only.
      // Note: label is not added through the translation files
      // to make filter customizations a bit easier.
      options: [
        { key: 'men', label: 'Men' },
        { key: 'women', label: 'Women' },
        { key: 'kids', label: 'Kids' },
      ],
    },
  },
  {
    id: 'amenities',
    label: 'Amenities',
    type: 'SelectMultipleFilter',
    group: 'secondary',
    queryParamNames: ['pub_amenities'],
    config: {
      // Schema type options: 'enum', 'multi-enum'
      // Both types can work so that user selects multiple values when filtering search results.
      // With "enum" the functionality will be OR-semantics (Nike OR Adidas OR Salomon)
      // With "multi-enum" it's possible to use both AND and OR semantics with searchMode config.
      schemaType: 'multi-enum',

      // Optional modes: 'has_all', 'has_any'
      // Note: this is relevant only for schema type 'multi-enum'
      // https://www.sharetribe.com/api-reference/marketplace.html#extended-data-filtering
      searchMode: 'has_all',

      // "key" is the option you see in Flex Console.
      // "label" is set here for this web app's UI only.
      // Note: label is not added through the translation files
      // to make filter customizations a bit easier.
      options: [
        { key: 'towels', label: 'Towels' },
        { key: 'bathroom', label: 'Bathroom' },
        { key: 'swimming_pool', label: 'Swimming pool' },
        { key: 'barbeque', label: 'Barbeque' },
      ],
    },
  },
];

describe('EditListingDetailsForm', () => {
  it('matches snapshot', () => {
    const tree = renderDeep(
      <EditListingDetailsForm
        intl={fakeIntl}
        dispatch={noop}
        onSubmit={v => v}
        saveActionMsg="Save details"
        updated={false}
        updateInProgress={false}
        disabled={false}
        ready={false}
        processInfos={processInfos}
        filterConfigs={filters}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
