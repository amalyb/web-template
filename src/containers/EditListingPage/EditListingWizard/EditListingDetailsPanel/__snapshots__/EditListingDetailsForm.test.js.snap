// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListingDetailsForm matches snapshot 1`] = `
<form
  action="/"
  className="root"
  method="post"
  onSubmit={[Function]}
>
  <div
    className="root title"
  >
    <label
      htmlFor="title"
    >
      EditListingDetailsForm.title
    </label>
    <input
      className="input"
      id="title"
      maxLength={60}
      name="title"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="EditListingDetailsForm.titlePlaceholder"
      type="text"
      value=""
    />
  </div>
  <div
    className="root description"
  >
    <label
      htmlFor="description"
    >
      EditListingDetailsForm.description
    </label>
    <textarea
      className="input textarea"
      id="description"
      maxLength={5000}
      name="description"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="EditListingDetailsForm.descriptionPlaceholder"
      rows={1}
      value=""
    />
  </div>
  <input
    name="transactionProcessAlias"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="hidden"
    value=""
  />
  <button
    className="root submitButton"
    disabled={true}
    type="submit"
  >
    Save details
  </button>
</form>
`;
